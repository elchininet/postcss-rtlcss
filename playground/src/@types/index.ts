export type NumberOrString = string | number;
export type NumberOrStringArray = NumberOrString[];

export interface ObjectProps<T> {
    [key: string]: T;
}