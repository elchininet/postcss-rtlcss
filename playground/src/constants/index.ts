export enum COLORS {
    action = '#18AEC9',
    active = '#026C7F',
    gray = '#DDD',
    gray_dark = '#555',
    gray_darkest = '#1E1E1E',
    gray_light = '#EFEFEF',
    white = '#FFF'
}