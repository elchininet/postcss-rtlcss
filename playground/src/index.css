html,
body {
    height: 100%;
    margin: 0;
    padding: 0;
}

body {
    font-family: Arial, Helvetica, sans-serif;
    overscroll-behavior-y: contain;
}

#app {
    height: 100%;
}