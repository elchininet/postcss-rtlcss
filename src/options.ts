import { Mode, Source } from '@types';

export { Mode, Source };