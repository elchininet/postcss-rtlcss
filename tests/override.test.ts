import postcss from 'postcss';
import { postcssRTLCSS, PluginOptions, Mode, Source } from '../src';
import { readCSSFile } from './test-utils';

const baseOptions: PluginOptions = {mode: Mode.override};

describe('Override Tests', (): void => {

  let input = '';

  beforeEach(async (): Promise<void> => {
    input = input || await readCSSFile('input.css');
  });

  it('Override Basic', (): void => {
    const options: PluginOptions = { ...baseOptions };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override {source: rtl}', (): void => {
    const options: PluginOptions = { ...baseOptions, source: Source.rtl };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override {processUrls: true}', (): void => {
    const options: PluginOptions = { ...baseOptions, processUrls: true };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override {processKeyFrames: true}', (): void => {
    const options: PluginOptions = { ...baseOptions, processKeyFrames: true };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override {source: rtl, processKeyFrames: true}', (): void => {
    const options: PluginOptions = { ...baseOptions, source: Source.rtl, processKeyFrames: true };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override {useCalc: true}', (): void => {
    const options: PluginOptions = { ...baseOptions, useCalc: true };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override custom ltrPrefix and rtlPrefix properties', (): void => {
    const options: PluginOptions = { ...baseOptions, ltrPrefix: '.ltr', rtlPrefix: '.rtl' };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override custom string map', (): void => {
    const stringMap: PluginOptions['stringMap'] = [
      {search: 'left', replace: 'right'},
      {search: 'ltr', replace: 'rtl'}
    ];
    const options: PluginOptions = { ...baseOptions, stringMap };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

  it('Override custom ltrPrefix and rtlPrefix properties as arrays', (): void => {
    const options: PluginOptions = { ...baseOptions, ltrPrefix: ['.ltr', '.left-to-right'], rtlPrefix: ['.rtl', '.right-to-left'] };
    const output = postcss([postcssRTLCSS(options)]).process(input);
    expect(output.css).toMatchSnapshot();
    expect(output.warnings()).toHaveLength(0);
  });

});