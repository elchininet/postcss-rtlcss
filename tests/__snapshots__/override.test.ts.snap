// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Override Tests Override {processUrls: true} 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

[dir=\\"rtl\\"] .test1, [dir=\\"rtl\\"] .test2 {
    background-image: url(\\"/folder/subfolder/icons/rtl/chevron-right.png\\");
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

[dir=\\"rtl\\"] .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

[dir=\\"rtl\\"] .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

[dir=\\"rtl\\"] .test5,
[dir=\\"rtl\\"] .test6,
[dir=\\"rtl\\"] .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

[dir=\\"rtl\\"] .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

[dir=\\"rtl\\"] .test9, [dir=\\"rtl\\"] .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

[dir=\\"rtl\\"] .test11:hover,
[dir=\\"rtl\\"] .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override {source: rtl} 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

[dir=\\"ltr\\"] .test1, [dir=\\"ltr\\"] .test2 {
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

[dir=\\"ltr\\"] .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

[dir=\\"ltr\\"] .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

[dir=\\"ltr\\"] .test5,
[dir=\\"ltr\\"] .test6,
[dir=\\"ltr\\"] .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

[dir=\\"ltr\\"] .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

[dir=\\"ltr\\"] .test9, [dir=\\"ltr\\"] .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

[dir=\\"ltr\\"] .test11:hover,
[dir=\\"ltr\\"] .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override {useCalc: true} 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

[dir=\\"rtl\\"] .test1, [dir=\\"rtl\\"] .test2 {
    background-position: calc(100% - 10px) 20px;
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

[dir=\\"rtl\\"] .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

[dir=\\"rtl\\"] .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
    transform-origin: calc(100% - 10px) 20px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

[dir=\\"rtl\\"] .test5,
[dir=\\"rtl\\"] .test6,
[dir=\\"rtl\\"] .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

[dir=\\"rtl\\"] .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

[dir=\\"rtl\\"] .test9, [dir=\\"rtl\\"] .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

[dir=\\"rtl\\"] .test11:hover,
[dir=\\"rtl\\"] .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override Basic 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

[dir=\\"rtl\\"] .test1, [dir=\\"rtl\\"] .test2 {
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

[dir=\\"rtl\\"] .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

[dir=\\"rtl\\"] .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

[dir=\\"rtl\\"] .test5,
[dir=\\"rtl\\"] .test6,
[dir=\\"rtl\\"] .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

[dir=\\"rtl\\"] .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

[dir=\\"rtl\\"] .test9, [dir=\\"rtl\\"] .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

[dir=\\"rtl\\"] .test11:hover,
[dir=\\"rtl\\"] .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override custom ltrPrefix and rtlPrefix properties 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

.rtl .test1, .rtl .test2 {
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

.rtl .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

.rtl .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

.rtl .test5,
.rtl .test6,
.rtl .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

.rtl .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

.rtl .test9, .rtl .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

.rtl .test11:hover,
.rtl .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override custom ltrPrefix and rtlPrefix properties as arrays 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

.rtl .test1, .right-to-left .test1, .rtl .test2, .right-to-left .test2 {
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

.rtl .test3, .right-to-left .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

.rtl .test4, .right-to-left .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

.rtl .test5,
.right-to-left .test5,
.rtl .test6,
.right-to-left .test6,
.rtl .test7,
.right-to-left .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

.rtl .test8, .right-to-left .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

.rtl .test9, .right-to-left .test9, .rtl .test10, .right-to-left .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

.rtl .test11:hover,
.right-to-left .test11:hover,
.rtl .test11:active,
.right-to-left .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;

exports[`Override Tests Override custom string map 1`] = `
".test1, .test2 {
    background-color: #FFF;
    background-position: 10px 20px;
    background-image: url(\\"/folder/subfolder/icons/ltr/chevron-left.png\\");
    border-radius: 0 2px 0 8px;
    color: #666;
    padding-right: 20px;
    text-align: left;
    transform: translate(-50%, 50%);
    width: 100%;
}

[dir=\\"rtl\\"] .test1, [dir=\\"rtl\\"] .test2 {
    border-radius: 2px 0 8px 0;
    padding-right: unset;
    padding-left: 20px;
    text-align: right;
    transform: translate(50%, 50%);
}

/* Comment not related to rtl */
.test3 {
    direction: ltr;
    margin: 1px 2px 3px;
    padding: 10px 20px;
    /* Property comment not related to rtl */
    text-align: center;
}

[dir=\\"rtl\\"] .test3 {
    direction: rtl;
}

.test4 {
    font-size: 10px;
    border-color: red;
    border-radius: 2px 4px 8px 16px;
    text-shadow: red 99px -1px 1px, blue 99px 2px 1px;
    transform-origin: 10px 20px;
    transform: scale(0.5, 0.5);
}

[dir=\\"rtl\\"] .test4 {
    border-radius: 4px 2px 16px 8px;
    text-shadow: red -99px -1px 1px, blue -99px 2px 1px;
}

.test5,
.test6,
.test7 {
    background: linear-gradient(0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border: 1px solid 2px;
    border-color: #666 #777 #888 #999;
    box-sizing: border-box;
    border-width: 1px 2px 3px 4px;
    position: absolute;
    left: 100px;
    transform: translateX(5px);
}

[dir=\\"rtl\\"] .test5,
[dir=\\"rtl\\"] .test6,
[dir=\\"rtl\\"] .test7 {
    background: linear-gradient(-0.25turn, #3F87A6, #EBF8E1, #F69D3C);
    border-color: #666 #999 #888 #777;
    border-width: 1px 4px 3px 2px;
    left: unset;
    right: 100px;
    transform: translateX(-5px);
}

/* rtl:novalid:ignore */
.test8 {
    background: linear-gradient(to left, #333, #333 50%, #EEE 75%, #333 75%);
    display: flex;
    padding-left: 10%;
}

[dir=\\"rtl\\"] .test8 {
    background: linear-gradient(to right, #333, #333 50%, #EEE 75%, #333 75%);
}

.test9, .test10 {
    background: linear-gradient(217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: 5px;
    padding: 0px 2px 3px 4px;
}

[dir=\\"rtl\\"] .test9, [dir=\\"rtl\\"] .test10 {
    background: linear-gradient(-217deg, rgba(255,0,0,.8), rgba(255,0,0,0) 70.71%),
                linear-gradient(-127deg, rgba(0,255,0,.8), rgba(0,255,0,0) 70.71%),
                linear-gradient(-336deg, rgba(0,0,255,.8), rgba(0,0,255,0) 70.71%);
    left: unset;
    right: 5px;
}

.test11:hover,
.test11:active {
    font-family: Arial, Helvetica;
    font-size: 20px;
    color: '#FFF';
    transform: translateY(10px);
    padding: 10px;
}

[dir=\\"rtl\\"] .test11:hover,
[dir=\\"rtl\\"] .test11:active {
    font-family: \\"Droid Arabic Kufi\\", Arial, Helvetica;
    font-size: 30px;
    color: #000;
    transform: translateY(10px) scaleX(-1);
    padding: 10px 20px;
}

#test12, #test13 {
    left: 10px;
    position: relative;
    text-align: right;
}

.test14 .test15 span {
    border-left-color: #777;
    margin: 10px 20px 30px 40px;
    transform: translate(100%, 10%);
}

.test16 {
    padding-right: 10px;
}

.test17 {
    animation: 5s flip 1s ease-in-out;
    padding: 10px 20px 40px 10px;
    text-align: right;
}

@keyframes flip {
    from {
        transform: translateX(-1000px);
    }
    to {
        transform: translateX(0);
    }
}"
`;
